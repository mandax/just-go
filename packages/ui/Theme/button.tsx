import * as React from "react";
import { fontMedium } from "../Theme/font";
import { shadow, Blur } from "../Theme/shadow";
import { container, roundedBorder } from "../Theme/container";
import { px, seconds, NumberToCSSUnit } from "../Theme/units";

import Theme from "../Theme";

const buttonBase = (
  hoverColor: string = Theme.COLOR_BASE,
  bgHoverColor: string = Theme.COLOR_BRAND,
  color: string = Theme.COLOR_PRIMARY,
  bgColor: string = Theme.COLOR_BASE,
) => (
  scale: NumberToCSSUnit,
  hover: boolean,
  bordered: boolean = true
): React.CSSProperties => {

  const animationSpeed = seconds(Theme.ANIMATION_SPEED);

  const hoverProps = hover && {
    ...shadow(),

    backgroundColor: bgHoverColor,
    borderColor: bgHoverColor,
    color: hoverColor,
    cursor: 'pointer'
  }

  return {
  ...container(1, 0.5),
  ...roundedBorder(Theme.BUTTON_BORDER_RADIUS),
  ...fontMedium(1, Theme.FONT_CONDENSED),
  ...shadow(Blur.Small),

  marginRight: scale(0.5),
  display: 'inline-block',
  fontSize: scale(Theme.BUTTON_FONT_SIZE),
  borderWidth: px(Theme.BUTTON_BORDER_SIZE),
  borderStyle: 'solid',
  borderColor: bordered ? color : bgColor,
  backgroundColor: bgColor,

  transition: `
    color ${animationSpeed} ease,
    border-color ${animationSpeed} ease,
    box-shadow ${animationSpeed} ease,
    background-color ${animationSpeed} ease
  `,

  ...hoverProps
  }
};

export const button = buttonBase();

export const buttonNeutro = buttonBase(Theme.COLOR_PRIMARY, Theme.COLOR_BASE_3);

export const buttonAccent = buttonBase(Theme.COLOR_BASE, Theme.COLOR_ACCENT);