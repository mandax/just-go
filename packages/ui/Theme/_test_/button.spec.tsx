import * as button from "../button";

describe('helper button', () => {
    
  it('should generate a default button', () => {
    expect(button.buttonBase()()).toMatchSnapshot()
  })

  it('should generate a new color button', () => {
    expect(button.buttonBase('blue', 'green', 'yellow', 'red')())
      .toMatchSnapshot();
  })

  it('should generate a button css with new scale function', () => {
    const scaleFunc = (a: number) => a.toString();
    expect(button.buttonBase()(scaleFunc)).toMatchSnapshot();
  })

  it('should generate a button css hovered', () => {
    expect(button.buttonBase()(undefined, true)).toMatchSnapshot();
  })

  it('should generate a button css without border', () => {
    expect(button.buttonBase()(undefined, undefined, false)).toMatchSnapshot();
  })

  it('should return a default css button', () => {
    expect(button.button()).toMatchSnapshot();
  })

  it('should return an accent css button', () => {
    expect(button.buttonAccent()).toMatchSnapshot();
  })

  it('should return a neutro css button', () => {
    expect(button.buttonNeutro()).toMatchSnapshot();
  })
})