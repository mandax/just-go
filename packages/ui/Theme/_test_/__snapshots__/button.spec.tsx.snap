// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper button should generate a button css hovered 1`] = `
Object {
  "backgroundColor": "#F4511E",
  "borderColor": "#F4511E",
  "borderRadius": 8,
  "borderStyle": "solid",
  "borderWidth": "1px",
  "boxShadow": "0px 0px 8px rgba(0, 0, 0, 0.1)",
  "color": "#FFFFFF",
  "cursor": "pointer",
  "display": "inline-block",
  "fontFamily": "Barlow Condensed",
  "fontSize": "1.2rem",
  "fontWeight": 500,
  "marginRight": "0.5rem",
  "padding": "0.5rem 1rem",
  "transition": "
      color 0.3s ease,
      border-color 0.3s ease,
      box-shadow 0.3s ease,
      background-color 0.3s ease
    ",
}
`;

exports[`helper button should generate a button css with new scale function 1`] = `
Object {
  "backgroundColor": "#FFFFFF",
  "borderColor": "#263238",
  "borderRadius": 8,
  "borderStyle": "solid",
  "borderWidth": "1px",
  "boxShadow": "0px 0px 4px rgba(0, 0, 0, 0.1)",
  "color": "#263238",
  "display": "inline-block",
  "fontFamily": "Barlow Condensed",
  "fontSize": "1.2",
  "fontWeight": 500,
  "marginRight": "0.5",
  "padding": "0.5rem 1rem",
  "transition": "
      color 0.3s ease,
      border-color 0.3s ease,
      box-shadow 0.3s ease,
      background-color 0.3s ease
    ",
}
`;

exports[`helper button should generate a button css without border 1`] = `
Object {
  "backgroundColor": "#FFFFFF",
  "borderColor": "#FFFFFF",
  "borderRadius": 8,
  "borderStyle": "solid",
  "borderWidth": "1px",
  "boxShadow": "0px 0px 4px rgba(0, 0, 0, 0.1)",
  "color": "#263238",
  "display": "inline-block",
  "fontFamily": "Barlow Condensed",
  "fontSize": "1.2rem",
  "fontWeight": 500,
  "marginRight": "0.5rem",
  "padding": "0.5rem 1rem",
  "transition": "
      color 0.3s ease,
      border-color 0.3s ease,
      box-shadow 0.3s ease,
      background-color 0.3s ease
    ",
}
`;

exports[`helper button should generate a default button 1`] = `
Object {
  "backgroundColor": "#FFFFFF",
  "borderColor": "#263238",
  "borderRadius": 8,
  "borderStyle": "solid",
  "borderWidth": "1px",
  "boxShadow": "0px 0px 4px rgba(0, 0, 0, 0.1)",
  "color": "#263238",
  "display": "inline-block",
  "fontFamily": "Barlow Condensed",
  "fontSize": "1.2rem",
  "fontWeight": 500,
  "marginRight": "0.5rem",
  "padding": "0.5rem 1rem",
  "transition": "
      color 0.3s ease,
      border-color 0.3s ease,
      box-shadow 0.3s ease,
      background-color 0.3s ease
    ",
}
`;

exports[`helper button should generate a new color button 1`] = `
Object {
  "backgroundColor": "red",
  "borderColor": "yellow",
  "borderRadius": 8,
  "borderStyle": "solid",
  "borderWidth": "1px",
  "boxShadow": "0px 0px 4px rgba(0, 0, 0, 0.1)",
  "color": "#263238",
  "display": "inline-block",
  "fontFamily": "Barlow Condensed",
  "fontSize": "1.2rem",
  "fontWeight": 500,
  "marginRight": "0.5rem",
  "padding": "0.5rem 1rem",
  "transition": "
      color 0.3s ease,
      border-color 0.3s ease,
      box-shadow 0.3s ease,
      background-color 0.3s ease
    ",
}
`;

exports[`helper button should return a default css button 1`] = `
Object {
  "backgroundColor": "#FFFFFF",
  "borderColor": "#263238",
  "borderRadius": 8,
  "borderStyle": "solid",
  "borderWidth": "1px",
  "boxShadow": "0px 0px 4px rgba(0, 0, 0, 0.1)",
  "color": "#263238",
  "display": "inline-block",
  "fontFamily": "Barlow Condensed",
  "fontSize": "1.2rem",
  "fontWeight": 500,
  "marginRight": "0.5rem",
  "padding": "0.5rem 1rem",
  "transition": "
      color 0.3s ease,
      border-color 0.3s ease,
      box-shadow 0.3s ease,
      background-color 0.3s ease
    ",
}
`;

exports[`helper button should return a neutro css button 1`] = `
Object {
  "backgroundColor": "#FFFFFF",
  "borderColor": "#263238",
  "borderRadius": 8,
  "borderStyle": "solid",
  "borderWidth": "1px",
  "boxShadow": "0px 0px 4px rgba(0, 0, 0, 0.1)",
  "color": "#263238",
  "display": "inline-block",
  "fontFamily": "Barlow Condensed",
  "fontSize": "1.2rem",
  "fontWeight": 500,
  "marginRight": "0.5rem",
  "padding": "0.5rem 1rem",
  "transition": "
      color 0.3s ease,
      border-color 0.3s ease,
      box-shadow 0.3s ease,
      background-color 0.3s ease
    ",
}
`;

exports[`helper button should return an accent css button 1`] = `
Object {
  "backgroundColor": "#FFFFFF",
  "borderColor": "#263238",
  "borderRadius": 8,
  "borderStyle": "solid",
  "borderWidth": "1px",
  "boxShadow": "0px 0px 4px rgba(0, 0, 0, 0.1)",
  "color": "#263238",
  "display": "inline-block",
  "fontFamily": "Barlow Condensed",
  "fontSize": "1.2rem",
  "fontWeight": 500,
  "marginRight": "0.5rem",
  "padding": "0.5rem 1rem",
  "transition": "
      color 0.3s ease,
      border-color 0.3s ease,
      box-shadow 0.3s ease,
      background-color 0.3s ease
    ",
}
`;
